backend/file.go

```diff
@@ -0,0 +1,237 @@
+func (d *FileStore) Get(_ context.Context, _ *Profile, key string) ([]byte, error) {
+	storeData, err := d.getStore()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, data := range storeData {
+		if data.KeyInfo.Name == key {
+			data, err := d.decrypt(data.Value, d.cryptoKey)
+			if err != nil {
+				return nil, fmt.Errorf("message authentication failed")
+			}
+			return data, nil
+		}
+	}
+	return nil, fmt.Errorf("%s not found", key)
+}
+
+// List reads the store from file, and fetch all keys
+func (d *FileStore) List(_ context.Context, _ *Profile) ([]Key, error) {
+	storeData, err := d.getStore()
+	if err != nil {
+		return nil, err
+	}
+	var keys []Key
+	for _, info := range storeData {
+		keys = append(keys, info.KeyInfo)
+	}
+	return keys, nil
+}
+
+// CheckExists checks if given key exists in the (file)store
+func (d *FileStore) CheckExists(_ context.Context, _ *Profile, key string) (bool, error) {
+	storeData, err := d.getStore()
+	if err != nil {
+		return false, err
+	}
+
+	for _, data := range storeData {
+		if data.KeyInfo.Name == key {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// Put a new Key with encrypted password in the store. Put overwrites the entire store file with the updated store
+func (d *FileStore) Put(_ context.Context, _ *Profile, key, value string) error {
+	if err := d.createStoreIfNotExists(); err != nil {
+		return err
+	}
+	encryptedData := d.encrypt([]byte(value), d.cryptoKey)
+	value = ""
+
+	owner := ""
+	currUser, err := user.Current()
+	if err == nil {
+		owner = currUser.Name
+	}
+	newStore := FileStoreEntry{
+		Value: encryptedData,
+		KeyInfo: Key{
+			Name:      key,
+			CreatedAt: time.Now(),
+			Owner:     owner,
+			Info:      "",
+		},
+	}
+
+	var store []FileStoreEntry
+	discStoreEntries, err := d.getStore()
+	if err != nil {
+		return err
+	}
+	if discStoreEntries != nil {
+		store = append(store, discStoreEntries...)
+	}
+	store = append(store, newStore)
+	data, err := json.Marshal(&store)
+	if err := ioutil.WriteFile(d.storeLocation, data, 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Delete a key from the store. Delete overwrites the entire store file with the updated store values
+func (d *FileStore) Delete(_ context.Context, _ *Profile, key string) error {
+	discStoreEntries, err := d.getStore()
+	if err != nil {
+		return err
+	}
+	var newDiscStore []FileStoreEntry
+	for _, entry := range discStoreEntries {
+		if entry.KeyInfo.Name != key {
+			newDiscStore = append(newDiscStore, entry)
+		}
+	}
+
+	data := []byte("")
+	// prevents "nil" being written to file
+	if len(newDiscStore) > 0 {
+		data, err = json.Marshal(&newDiscStore)
+	}
+	if err := ioutil.WriteFile(d.storeLocation, data, 0600); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *FileStore) Close() error {
+	return nil
+}
+
+// encrypt data based on the argon2 hashing algorithm and xchacha20 cipher algorithm
+func (d *FileStore) encrypt(data, pass []byte) []byte {
```

Comment:
would be good to have tests for the "core" of this backend.

Reply:
Added some basic tests to prove the encrypt/decrypt works as expected, so this is probably not adhering to your expectation yet. But it requires some refactoring of the code such that we don't have ioutil.x references in the core methods (get, put, list, delete).

Will look into this.